src/scripts/main.js

```diff
@@ -1,3 +1,229 @@
 'use strict';
 
-// write your code here
+const table = document.querySelector('tbody');
+const button = document.querySelector('.button');
+const startMessage = document.querySelector('.message-start');
+const loseMessage = document.querySelector('.message-lose');
+const winMessage = document.querySelector('.message-win');
+const gameScore = document.querySelector('.game-score');
+
+const rowsCount = 4;
+const colsCount = 4;
+let field;
+let score;
+
+button.addEventListener('click', () => {
+  if (button.classList.contains('start')) {
+    button.classList.remove('start');
+    button.classList.add('restart');
+    button.textContent = 'Restart';
+    startMessage.classList.add('hidden');
+  } else {
+    loseMessage.classList.add('hidden');
+    winMessage.classList.add('hidden');
+  }
+
+  field = [
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+  ];
+
+  score = 0;
+  gameScore.textContent = score;
+
+  addRandomCell();
+  addRandomCell();
+});
+
+document.addEventListener('keydown', (e) => {
+  const oldField = deepCopyArray(field);
+
+  if (
+    !isMoved(moveUp(oldField))
+    && !isMoved(moveDown(oldField))
+    && !isMoved(moveRight(oldField))
+    && !isMoved(moveLeft(oldField))
+  ) {
+    loseMessage.classList.remove('hidden');
+
+    return false;
+  }
+
+  switch (e.key) {
+    case 'ArrowUp':
+      field = deepCopyArray(moveUp(field));
+      break;
+    case 'ArrowDown':
+      field = deepCopyArray(moveDown(field));
+      break;
+    case 'ArrowRight':
+      field = deepCopyArray(moveRight(field));
+      break;
+    case 'ArrowLeft':
+      field = deepCopyArray(moveLeft(field));
+      break;
+  }
+
+  if (isMoved(oldField)) {
+    addRandomCell();
+    gameScore.textContent = `${score}`;
+  }
+
+  if (isWin()) {
+    winMessage.classList.remove('hidden');
+  }
+});
+
+function isWin() {
+  for (let i = 0; i < rowsCount; i++) {
+    for (let j = 0; j < colsCount; j++) {
+      if (field[i][j] === 2048) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+function deepCopyArray(arr) {
+  return JSON.parse(JSON.stringify(arr));
+}
+
+function isMoved(oldField) {
+  for (let i = 0; i < rowsCount; i++) {
+    for (let j = 0; j < colsCount; j++) {
+      if (oldField[i][j] !== field[i][j]) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+function moveUp(inputField) {
+  const movedField = deepCopyArray(inputField);
+
+  for (let i = 0; i < rowsCount; i++) {
+    let newColumn = [
+      movedField[0][i],
+      movedField[1][i],
+      movedField[2][i],
+      movedField[3][i],
+    ];
+
+    newColumn = move(newColumn.reverse()).reverse();
+
+    for (let j = 0; j < newColumn.length; j++) {
+      movedField[j][i] = newColumn[j];
+    }
+  }
+
+  return movedField;
+}
+
+function moveDown(inputField) {
+  const movedField = deepCopyArray(inputField);
+
+  for (let i = 0; i < rowsCount; i++) {
+    let newColumn = [
+      movedField[0][i],
```

Comment:
You are duplicating code in `moveUp`, `moveDown`, `moveRight` and `moveLeft` functions. Try to create a helper function to avoid code duplication.